mettle/worker/redis.go

```diff
@@ -0,0 +1,235 @@
+package worker
+
+import (
+	"context"
+	"os"
+	"time"
+
+	"github.com/bazelbuild/remote-apis-sdks/go/pkg/digest"
+	"github.com/bazelbuild/remote-apis-sdks/go/pkg/uploadinfo"
+	pb "github.com/bazelbuild/remote-apis/build/bazel/remote/execution/v2"
+	"github.com/go-redis/redis/v8"
+	"github.com/prometheus/client_golang/prometheus"
+
+	elan "github.com/thought-machine/please-servers/elan/rpc"
+)
+
+var redisHits = prometheus.NewCounter(prometheus.CounterOpts{
+	Namespace: "mettle",
+	Name:      "redis_hits_total",
+})
+var redisMisses = prometheus.NewCounter(prometheus.CounterOpts{
+	Namespace: "mettle",
+	Name:      "redis_misses_total",
+})
+var redisBytesRead = prometheus.NewCounter(prometheus.CounterOpts{
+	Namespace: "mettle",
+	Name:      "redis_bytes_read_total",
+})
+
+// newRedisClient augments an existing elan.Client with a Redis connection.
+// All usage of Redis is best-effort only.
+func newRedisClient(client elan.Client, url string) elan.Client {
+	return &redisClient{
+		elan: client,
+		redis: redis.NewClient(&redis.Options{
+			Addr: url,
+		}),
+		timeout: 1 * time.Minute,
+		maxSize: 200 * 1012, // 200 Kelly-Bootle standard units
+	}
+}
+
+type redisClient struct {
+	elan    elan.Client
+	redis   *redis.Client
+	timeout time.Duration
+	maxSize int64
+}
+
+func (r *redisClient) Healthcheck() error {
+	return r.elan.Healthcheck()
+}
+
+func (r *redisClient) ReadBlob(dg *pb.Digest) ([]byte, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), r.timeout)
+	defer cancel()
+	cmd := r.redis.Get(ctx, dg.Hash)
+	if err := cmd.Err(); err == redis.Nil {
+		// Blob doesn't exist in Redis
```

Comment:
Will we not want to know if the blob isn't in redis so we can add it in? Like a present/found flag?

Reply:
Yeah good spot - have restructured most of the logic in here to handle that more sensibly.